src/modules/plotter.py

```diff
@@ -0,0 +1,134 @@
+        self.mode = mode
+
+        with open(self.metrics_path, 'r') as f:
+            self.data = json.load(f)
+
+        # Extract dimensions for metrics
+        self.num_repeats = len(set(entry['repeat'] for entry in self.data[self.mode]))
+        self.num_episodes = len(set(entry['episode'] for entry in self.data[self.mode]))
+        self.num_agents = len(self.data[self.mode][0]['actions']) if 'actions' in self.data[self.mode][0] else 0
+        self.joint_action_size = len(self.data[self.mode][0]['joint action frequencies']) if 'joint action frequencies' in self.data[self.mode][0] else 0
+
+    def extract_metrics(self):
+        """
+        Extract all relevant metrics from the data into numpy arrays for easy processing.
+
+        Returns:
+        - A dictionary with metrics in numpy array format.
+        """
+        metrics = {
+            'agent_metrics': {},
+            'cumulative_return': None,
+            'joint_action_frequencies': None
+        }
+
+        # Initialize arrays for storing metrics
+        for key in self.data[self.mode][0].keys():
+            if key not in ['repeat', 'episode', 'actions', 'rewards', 'joint action frequencies', 'metrics']:
+                metrics['agent_metrics'][key] = np.zeros((self.num_repeats, self.num_episodes, self.num_agents))
+
+        metrics['joint_action_frequencies'] = np.zeros((self.num_repeats, self.num_episodes, self.joint_action_size))
+        metrics['cumulative_return'] = np.zeros((self.num_repeats, self.num_episodes)) if self.mode == 'evaluation' else None
+
+        # Fill metrics
+        for entry in self.data[self.mode]:
+            repeat = entry['repeat']
+            episode = entry['episode']
+
+            for key, value in entry.items():
+                if key in metrics['agent_metrics']:
+                    for agent_idx, agent_value in enumerate(value):
+                        metrics['agent_metrics'][key][repeat, episode, agent_idx] = agent_value
+                elif key == 'joint action frequencies':
+                    metrics['joint_action_frequencies'][repeat, episode] = value
+                elif key == 'metrics':
+                    for metric_key, metric_value in value['metrics'].items():
+                        if isinstance(metric_value, list):
+                            if metric_key not in metrics['agent_metrics']:
+                                metrics['agent_metrics'][metric_key] = np.zeros((self.num_repeats, self.num_episodes, self.num_agents))
+                            for agent_idx, agent_value in enumerate(metric_value):
+                                metrics['agent_metrics'][metric_key][repeat, episode, agent_idx] = agent_value
+                        elif metric_key == 'cumulative return':
+                            metrics['cumulative_return'][repeat, episode] = metric_value
+
+        return metrics
+
+    def plot_metric(self, metric_data, metric_name, xlabel='Episodes', ylabel=None):
```

Comment:
Find a concensus on graphical settings for graphs, and check outliers in CI, and change nomenclature in joint action frquencies graphs (solid lines should only be for agent specific metrics, something else is needed for JAF defect, and even cooperate because -- lines look weird on roller costers).